### DISCUSSION TITLE
Tesla Fleet API enrolment

### DISCUSSION THREAD

**Discussion by RichieB2B** [ngardiner/TWCManager#570] [General]
This is a continued discussion from #546

Here is a nice tutorial that can be used to register TWCManager as a personal app with Tesla: https://shankar-k.medium.com/tesla-developer-api-guide-account-setup-app-creation-registration-and-third-party-da24aba1bddd

@ChutneyMary Are you still having issue getting your app registered? At which step do you get stuck following the above guide?

**Comment by danielsan1** [comment]
Thanks for the link! Beginner question: The requirements for the guide are a domain name and a webserver running on my TWCManager Raspberry Pi, right? (or any other machine?)

**Reply by RichieB2B** [reply]
Any other machine is fine. It is just used for hosting the certificate of your app and for the callbacks after your app has been added to a car.

**Comment by MikeBishop** [comment]
As an additional resource, [this repo](https://github.com/MikeBishop/fleetapi-tokens) aims to simplify what needs to be hosted on that web server into a Docker container. Still need the domain, however.

**Comment by ChutneyMary** [comment]
@RichieB2B - I get stuck at the app registration step with a message 'Unable to fulfill request. Please try again later'.

The specific step is after my domain name is entered (including callback URL), then the API Scopes are checked as required. The next button results in the message above.

I did have some progresss when an option was provided - 1. Machine to Machine or  2. 'Authorization Code and Machine-to-Machine'.

The first option 'Machine to Machine' provides a Client ID and Client Secret. But no option to provide my Domain details.

I have the com.tesla.3p.public-key.pem available for retrieval on my public facing domain address. But it's doing nothing without the proper app registration success.

**Reply by RichieB2B** [reply]
I had the same kind of issue. When I checked the logs of my website I noticed that the page `/.well-known/appspecific/com.tesla.3p.public-key.pem` could not be retrieved. This caused the app registration to fail. After I fixed this page I was able to finish the registration.

**Reply by ChutneyMary** [reply]
I'm testing it on my phone with wifi turned off and connected to the mobile network. I navigate to -
`https://myhostname.ddns.net/.well-known/appspecific/com.tesla.3p.public-key.pem`
The certificate downloads fine. 
I've tried a different free domain through noip.com without success. I'm beginning to wonder whether it's the domain ddns.net that's the issue. Can anyone suggest working free domains that Tesla accepts?
I've tried a dozen different app names and app descriptions to ensure I'm not treading on someone else. Tesla is not exactly helpful when it comes to addressing issues like this...

**Reply by RichieB2B** [reply]
Tesla registers the apps on the domain level, not FQDN. So in your example "ddns.net" which is a dynamic DNS provider. Even if Tesla would allow this (they probably blacklisted it) the chance is someone else already registered an app using "ddns.net". You need to use a domain you actually own.

**Reply by ChutneyMary** [reply]
Thanks. My understanding of what 'owning' the domain means was deficient. That explanation helps. 
Now I just need to decide whether I'm happy to fork out money for a hosting service to retrieve my own data for my own vehicle....grrrr.
And this is before Tesla even starts charging for the Fleet API usage.
Hopefully someone can get the [Bluetooth integration](https://github.com/ngardiner/TWCManager/discussions/562) working so TWCManager can be a truly off-line system (wishes he studied coding in school).

**Comment by MikeBishop** [comment]
In order to use the Bluetooth integration, you have to have registered on the portal, so all of these roadblocks are still there. The key that Tesla is fetching from that URL is used to authenticate Bluetooth commands, IIUC.

**Reply by ChutneyMary** [reply]
I interpret that differently based on what I read in the Developer docs -

https://shankar-k.medium.com/tesla-developer-api-guide-ble-key-pair-auth-and-vehicle-commands-part-3-485e4a357e7d

https://github.com/teslamotors/vehicle-command/tree/main/cmd/tesla-control

"Sending commands
You should now be able to send commands over BLE:

tesla-control -ble lock
If you've set up your OAuth token (see [repository README file](https://github.com/teslamotors/vehicle-command/blob/main/README.md)), you can also send commands over the Internet:

The words that stand out are **If you've set up your OAuth token** you can also send commands over the Internet." ie. internet authentication is not mandatory???

**Reply by MikeBishop** [reply]
Ah, you appear to be correct. The BLE code can be enrolled by using the keycard to approve it while physically in the car. That is promising!

**Reply by MikeBishop** [reply]
Since I was moving over to Fleet API proxy today, I tried getting the BLE working from the command line. I consistently got this:
```
Failed to execute command: MESSAGEFAULT_ERROR_INSUFFICIENT_PRIVILEGES
```

The Bluetooth tools in the OS seem to work, so I'm not sure what's up with that.

**Comment by alexeiw123** [comment]
@RichieB2B could you help with the final and crucial stage of getting TWCManager to use the fleetAPI, once I've got all the other steps sorted out? I assume somewhere I need to give it the appropriate addresses.

I have the cert hosted, call back etc. I did all this from the home assistant integration, which is working just fine now. I assume that I can use the same cert/app etc. for TWC Manager (hopefully teslamate as well). Any help appreciated. Manually starting the charge is feeling very old fashioned after using TWCManager without issue for so long!

**Reply by RichieB2B** [reply]
So you used [this guide](https://github.com/llamafilm/tesla-http-proxy-addon/blob/main/tesla_http_proxy/DOCS.md) to set up the Tesla http proxy? You can use this to generate a new refresh token for TWC Manager, see the "Generate token from URL" step in the guide. Copy-paste the generated refresh token to the TWCManager settings page, or manually add it to `/etc/twcmanager/settings.json`

Stop TWCManager

In your `/etc/twcmanager/config.json` add the following lines, using the values that fit your local setup:
```
"teslaApiClientID": "client-id-of-registered-app",
"teslaProxy": "https://localhost:4443",
"teslaProxyCert": "/path/to/public_key.pem"
```
See [here](https://github.com/ngardiner/TWCManager/blob/2a6a914657e5414ef96987b1184333c77bc22e75/etc/twcmanager/config.json#L134-L155) for more documentation about these config options.

Start TWCManager

**Reply by alexeiw123** [reply]
Thanks, yes that is what I've used - I will give this a go. I didn't realise the UI would accept the Tesla-api refresh token.

> See [here](https://github.com/ngardiner/TWCManager/blob/2a6a914657e5414ef96987b1184333c77bc22e75/etc/twcmanager/config.json#L134-L155) for more documentation about these config options.

I had seen this added in git but after updating TWC Manager those commented lines weren't in my config.json, so wasn't sure if I'd missed something. Over the next day or so I'll get some time to try this, will report back.

**Reply by alexeiw123** [reply]
Well, I generated another refresh token and added it to TWCManager, then added the following lines to what I thought they would need to be. As I'm using a [cloudflare tunne](https://github.com/llamafilm/tesla-http-proxy-addon/wiki/Cloudflare-Argo-Tunnel) for my HA instance, I'm basically unsure what I should use for `teslaProxy`.

All I've managed to achieve so far is TWCManager running but the 'home' tab doesn't load. settings etc. does.
Once I comment out the below lines (redacted my domain etc.), it works again with no API functionality. I've tried a few different variations but honestly I don't understand it well enough to be doing anything other than having a punt.

```
        "teslaApiClientID": "<snip>",
        "teslaProxy": "https://my.domain.xyz",
        "teslaProxyCert": "/.well-known/appspecific/com.tesla.3p.public-key.pem",
```

Edit: OK even more confused now - I commented out the above lines and now the api is _working_. I guess something in the config I tried allowed for the refresh token to authorise. it's working now but I suspect it won't refresh at some point in the future.

**Reply by RichieB2B** [reply]
git will never magically change your active config files like `/etc/twcmanager/config.json` but only the files in your local copy of the source code repository like `TWCManager/etc/twcmanager/config.json`

I am not familiar with "[cloudflare tunnel](https://github.com/llamafilm/tesla-http-proxy-addon/wiki/Cloudflare-Argo-Tunnel)" but reading that documentation I think you need to use the URL from step 5 for `teslaProxy` and the `active universal certificate` from step 6 for `teslaProxyCert`. You will need to **copy** this in PEM format to your TWCManager machine though. `teslaProxyCert` cannot be a remote location. For example you can save it as `/etc/twcmanager/proxy.pem` and then use:

```
"teslaProxyCert": "/etc/twcmanager/proxy.pem",
```

> Edit: OK even more confused now - I commented out the above lines and now the api is working. I guess something in the config I tried allowed for the refresh token to authorise. it's working now but I suspect it won't refresh at some point in the future. 

So you are using a Fleet API refresh token but commented out the proxy settings. That could mean one of two things:
1. You have an older Tesla that still supports direct commands
2. Only commands require the use of the Tesla Vehicle Command SDK and as a result using a Tesla HTTP Proxy. TWCManager status calls will work just fine without the proxy but you won't be able to start or stop a charge via the API.

In situation 1 you do not need a proxy at all. In situation 2 you can set the [Stop Charging Method](https://github.com/ngardiner/TWCManager/blob/2a6a914657e5414ef96987b1184333c77bc22e75/docs/Settings.md#stop-charging-method) to "Stop Responding to Slaves" or fix your Tesla HTTP Proxy setup.

In order for the Fleet API token to be refreshed you do need the provide the correct `teslaApiClientID` though:

```
"teslaApiClientID": "<snip>",
```

**Reply by alexeiw123** [reply]
> git will never magically change your active config files like `/etc/twcmanager/config.json` but only the files in your local copy of the source code repository like `TWCManager/etc/twcmanager/config.json`

Thankyou, this makes perfect sense of course.

> use the URL from step 5 for `teslaProxy`

Awesome, easy

> and the `active universal certificate` from step 6 for `teslaProxyCert`. You will need to **copy** this in PEM format to your TWCManager machine though.

OK this part is a bit more of a struggle. So the cert I'm after is not the Tesla API specific `/.well-known/appspecific/com.tesla.3p.public-key.pem`? It's an SSL certificate for the domain? My certs are created using let's encrypt but won't the expire if I copy them off to the twcmanager location?

> 1. You have an older Tesla that still supports direct commands
> 2. Only commands require the use of the Tesla Vehicle Command SDK and as a result using a Tesla HTTP Proxy. TWCManager status calls will work just fine without the proxy but you won't be able to start or stop a charge via the API.
> 
> In situation 1 you do not need a proxy at all. In situation 2 you can set the [Stop Charging Method](https://github.com/ngardiner/TWCManager/blob/2a6a914657e5414ef96987b1184333c77bc22e75/docs/Settings.md#stop-charging-method) to "Stop Responding to Slaves" or fix your Tesla HTTP Proxy setup.

Very interesting. My car is 2019 so almost certainly yes for 1. (and teslamate is still working fine on legacy API for now)
I tested with regards to 2. and while I can control the charge rate down to 2A using twcmanager (API is the only way to achieve this), You are right that I cannot start/stop the charge.

> In order for the Fleet API token to be refreshed you do need the provide the correct `teslaApiClientID` though:

I do have this detail right, I'm pretty sure!

I will continue to dabble using your advice, thankyou.

**Reply by alexeiw123** [reply]
I've run out of time again on this without success, here's my twcmanager log when I click on the 'home' tab in the UI (which still just loads a blank page). It appears that the teslaProxy address or the cert file that I've copied across aren't correct. Thanks for your help so far!

```
journalctl -f
-- Logs begin at Sat 2024-04-06 10:17:01 AEDT. --
Apr 19 08:45:38 TWCManager sudo[18741]: pam_unix(sudo:session): session opened for user root by pi(uid=0)
Apr 19 08:46:11 TWCManager sudo[18741]: pam_unix(sudo:session): session closed for user root
Apr 19 08:46:14 TWCManager sudo[18743]:       pi : TTY=pts/0 ; PWD=/etc/twcmanager ; USER=root ; COMMAND=/usr/sbin/service twcmanager restart
Apr 19 08:46:14 TWCManager sudo[18743]: pam_unix(sudo:session): session opened for user root by pi(uid=0)
Apr 19 08:46:15 TWCManager systemd[1]: Stopping Tesla Wall Charger Manager...
Apr 19 08:46:15 TWCManager systemd[1]: twcmanager.service: Main process exited, code=killed, status=15/TERM
Apr 19 08:46:15 TWCManager systemd[1]: twcmanager.service: Succeeded.
Apr 19 08:46:15 TWCManager systemd[1]: Stopped Tesla Wall Charger Manager.
Apr 19 08:46:15 TWCManager systemd[1]: Started Tesla Wall Charger Manager.
Apr 19 08:46:15 TWCManager sudo[18743]: pam_unix(sudo:session): session closed for user root
Apr 19 08:46:34 TWCManager python3[18752]: ----------------------------------------
Apr 19 08:46:34 TWCManager python3[18752]: Exception happened during processing of request from ('192.168.1.141', 56426)Apr 19 08:46:34 TWCManager python3[18752]: Traceback (most recent call last):
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 600, in urlopen
Apr 19 08:46:34 TWCManager python3[18752]:     chunked=chunked)
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 343, in _make_request
Apr 19 08:46:34 TWCManager python3[18752]:     self._validate_conn(conn)
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 841, in _validate_conn
Apr 19 08:46:34 TWCManager python3[18752]:     conn.connect()
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3/dist-packages/urllib3/connection.py", line 344, in connect
Apr 19 08:46:34 TWCManager python3[18752]:     ssl_context=context)
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3/dist-packages/urllib3/util/ssl_.py", line 344, in ssl_wrap_socket
Apr 19 08:46:34 TWCManager python3[18752]:     return context.wrap_socket(sock, server_hostname=server_hostname)
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3.7/ssl.py", line 412, in wrap_socket
Apr 19 08:46:34 TWCManager python3[18752]:     session=session
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3.7/ssl.py", line 853, in _create
Apr 19 08:46:34 TWCManager python3[18752]:     self.do_handshake()
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3.7/ssl.py", line 1117, in do_handshake
Apr 19 08:46:34 TWCManager python3[18752]:     self._sslobj.do_handshake()
Apr 19 08:46:34 TWCManager python3[18752]: ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)
Apr 19 08:46:34 TWCManager python3[18752]: During handling of the above exception, another exception occurred:
Apr 19 08:46:34 TWCManager python3[18752]: Traceback (most recent call last):
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/local/lib/python3.7/dist-packages/requests-2.25.1-py3.7.egg/requests/adapters.py", line 449, in send
Apr 19 08:46:34 TWCManager python3[18752]:     timeout=timeout
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 638, in urlopen
Apr 19 08:46:34 TWCManager python3[18752]:     _stacktrace=sys.exc_info()[2])
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3/dist-packages/urllib3/util/retry.py", line 398, in increment
Apr 19 08:46:34 TWCManager python3[18752]:     raise MaxRetryError(_pool, url, error or ResponseError(cause))
Apr 19 08:46:34 TWCManager python3[18752]: urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='my.secretdomain.com', port=443): Max retries exceeded with url: /api/1/vehicles (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)')))
Apr 19 08:46:34 TWCManager python3[18752]: During handling of the above exception, another exception occurred:
Apr 19 08:46:34 TWCManager python3[18752]: Traceback (most recent call last):
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/local/lib/python3.7/dist-packages/TWCManager/Vehicle/TeslaAPI.py", line 247, in car_api_available
Apr 19 08:46:34 TWCManager python3[18752]:     req = requests.get(url, headers=headers, verify=self.verifyCert)
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/local/lib/python3.7/dist-packages/requests-2.25.1-py3.7.egg/requests/api.py", line 76, in get
Apr 19 08:46:34 TWCManager python3[18752]:     return request('get', url, params=params, **kwargs)
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/local/lib/python3.7/dist-packages/requests-2.25.1-py3.7.egg/requests/api.py", line 61, in request
Apr 19 08:46:34 TWCManager python3[18752]:     return session.request(method=method, url=url, **kwargs)
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/local/lib/python3.7/dist-packages/requests-2.25.1-py3.7.egg/requests/sessions.py", line 542, in request
Apr 19 08:46:34 TWCManager python3[18752]:     resp = self.send(prep, **send_kwargs)
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/local/lib/python3.7/dist-packages/requests-2.25.1-py3.7.egg/requests/sessions.py", line 655, in send
Apr 19 08:46:34 TWCManager python3[18752]:     r = adapter.send(request, **kwargs)
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/local/lib/python3.7/dist-packages/requests-2.25.1-py3.7.egg/requests/adapters.py", line 514, in send
Apr 19 08:46:34 TWCManager python3[18752]:     raise SSLError(e, request=request)
Apr 19 08:46:34 TWCManager python3[18752]: requests.exceptions.SSLError: HTTPSConnectionPool(host='my.secretdomain.com', port=443): Max retries exceeded with url: /api/1/vehicles (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)')))
Apr 19 08:46:34 TWCManager python3[18752]: During handling of the above exception, another exception occurred:
Apr 19 08:46:34 TWCManager python3[18752]: Traceback (most recent call last):
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3.7/socketserver.py", line 650, in process_request_thread
Apr 19 08:46:34 TWCManager python3[18752]:     self.finish_request(request, client_address)
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3.7/socketserver.py", line 360, in finish_request
Apr 19 08:46:34 TWCManager python3[18752]:     self.RequestHandlerClass(request, client_address, self)
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/local/lib/python3.7/dist-packages/TWCManager/Control/HTTPControl.py", line 158, in __init__
Apr 19 08:46:34 TWCManager python3[18752]:     BaseHTTPRequestHandler.__init__(self, *args, **kwargs)
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3.7/socketserver.py", line 720, in __init__
Apr 19 08:46:34 TWCManager python3[18752]:     self.handle()
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3.7/http/server.py", line 426, in handle
Apr 19 08:46:34 TWCManager python3[18752]:     self.handle_one_request()
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/lib/python3.7/http/server.py", line 414, in handle_one_request
Apr 19 08:46:34 TWCManager python3[18752]:     method()
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/local/lib/python3.7/dist-packages/TWCManager/Control/HTTPControl.py", line 768, in do_GET
Apr 19 08:46:34 TWCManager python3[18752]:     "TeslaAPI"
Apr 19 08:46:34 TWCManager python3[18752]:   File "/usr/local/lib/python3.7/dist-packages/TWCManager/Vehicle/TeslaAPI.py", line 252, in car_api_available
Apr 19 08:46:34 TWCManager python3[18752]:     if req:
Apr 19 08:46:35 TWCManager python3[18752]: UnboundLocalError: local variable 'req' referenced before assignment
Apr 19 08:46:35 TWCManager python3[18752]: ----------------------------------------
```

**Reply by alexeiw123** [reply]
OK - one last dump.

I believe I have the address right now and I'm no longer getting an SSL error. I still get a blank landing page for TWCManager though:

```
journalctl -f
-- Logs begin at Fri 2024-04-19 09:19:47 AEST. --
Apr 19 09:49:16 TWCManager sudo[21252]: pam_unix(sudo:session): session opened for user root by pi(uid=0)
Apr 19 09:49:58 TWCManager sudo[21252]: pam_unix(sudo:session): session closed for user root
Apr 19 09:50:01 TWCManager sudo[21276]:       pi : TTY=pts/0 ; PWD=/etc/twcmanager ; USER=root ; COMMAND=/usr/sbin/service twcmanager restart
Apr 19 09:50:01 TWCManager sudo[21276]: pam_unix(sudo:session): session opened for user root by pi(uid=0)
Apr 19 09:50:02 TWCManager systemd[1]: Stopping Tesla Wall Charger Manager...
Apr 19 09:50:02 TWCManager systemd[1]: twcmanager.service: Main process exited, code=killed, status=15/TERM
Apr 19 09:50:02 TWCManager systemd[1]: twcmanager.service: Succeeded.
Apr 19 09:50:02 TWCManager systemd[1]: Stopped Tesla Wall Charger Manager.
Apr 19 09:50:02 TWCManager systemd[1]: Started Tesla Wall Charger Manager.
Apr 19 09:50:02 TWCManager sudo[21276]: pam_unix(sudo:session): session closed for user root
Apr 19 09:50:13 TWCManager python3[21282]: ----------------------------------------
Apr 19 09:50:13 TWCManager python3[21282]: Exception happened during processing of request from ('192.168.1.141', 58657)
Apr 19 09:50:13 TWCManager python3[21282]: Traceback (most recent call last):
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/local/lib/python3.7/dist-packages/TWCManager/Vehicle/TeslaAPI.py", line 247, in car_api_available
Apr 19 09:50:13 TWCManager python3[21282]:     req = requests.get(url, headers=headers, verify=self.verifyCert)
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/local/lib/python3.7/dist-packages/requests-2.25.1-py3.7.egg/requests/api.py", line 76, in get
Apr 19 09:50:13 TWCManager python3[21282]:     return request('get', url, params=params, **kwargs)
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/local/lib/python3.7/dist-packages/requests-2.25.1-py3.7.egg/requests/api.py", line 61, in request
Apr 19 09:50:13 TWCManager python3[21282]:     return session.request(method=method, url=url, **kwargs)
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/local/lib/python3.7/dist-packages/requests-2.25.1-py3.7.egg/requests/sessions.py", line 542, in request
Apr 19 09:50:13 TWCManager python3[21282]:     resp = self.send(prep, **send_kwargs)
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/local/lib/python3.7/dist-packages/requests-2.25.1-py3.7.egg/requests/sessions.py", line 649, in send
Apr 19 09:50:13 TWCManager python3[21282]:     adapter = self.get_adapter(url=request.url)
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/local/lib/python3.7/dist-packages/requests-2.25.1-py3.7.egg/requests/sessions.py", line 742, in get_adapter
Apr 19 09:50:13 TWCManager python3[21282]:     raise InvalidSchema("No connection adapters were found for {!r}".format(url))
Apr 19 09:50:13 TWCManager python3[21282]: requests.exceptions.InvalidSchema: No connection adapters were found for 'my.secretdomain.com:4430/api/1/vehicles'
Apr 19 09:50:13 TWCManager python3[21282]: During handling of the above exception, another exception occurred:
Apr 19 09:50:13 TWCManager python3[21282]: Traceback (most recent call last):
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/lib/python3.7/socketserver.py", line 650, in process_request_thread
Apr 19 09:50:13 TWCManager python3[21282]:     self.finish_request(request, client_address)
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/lib/python3.7/socketserver.py", line 360, in finish_request
Apr 19 09:50:13 TWCManager python3[21282]:     self.RequestHandlerClass(request, client_address, self)
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/local/lib/python3.7/dist-packages/TWCManager/Control/HTTPControl.py", line 158, in __init__
Apr 19 09:50:13 TWCManager python3[21282]:     BaseHTTPRequestHandler.__init__(self, *args, **kwargs)
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/lib/python3.7/socketserver.py", line 720, in __init__
Apr 19 09:50:13 TWCManager python3[21282]:     self.handle()
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/lib/python3.7/http/server.py", line 426, in handle
Apr 19 09:50:13 TWCManager python3[21282]:     self.handle_one_request()
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/lib/python3.7/http/server.py", line 414, in handle_one_request
Apr 19 09:50:13 TWCManager python3[21282]:     method()
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/local/lib/python3.7/dist-packages/TWCManager/Control/HTTPControl.py", line 768, in do_GET
Apr 19 09:50:13 TWCManager python3[21282]:     "TeslaAPI"
Apr 19 09:50:13 TWCManager python3[21282]:   File "/usr/local/lib/python3.7/dist-packages/TWCManager/Vehicle/TeslaAPI.py", line 252, in car_api_available
Apr 19 09:50:13 TWCManager python3[21282]:     if req:
Apr 19 09:50:13 TWCManager python3[21282]: UnboundLocalError: local variable 'req' referenced before assignment
Apr 19 09:50:13 TWCManager python3[21282]: ----------------------------------------
```

**Reply by RichieB2B** [reply]
> Very interesting. My car is 2019 so almost certainly yes for 1. (and teslamate is still working fine on legacy API for now)

So stop trying to get the proxy to work since it looks you don't need it. Simply only define `teslaApiClientID` and use the refresh token from the Fleet API (HA integration). Comment out `teslaProxy` and `teslaProxyCert` in `/etc/twcmanager/config.json`

**Only read on if you want to continue with getting the proxy to work.**

> My certs are created using let's encrypt but won't the expire if I copy them off to the twcmanager location?

It is very uncommon to use a public domain and an officially signed certificate for your Tesla HTTP Proxy. Generally the proxy is run on localhost with a self signed certificate that is valid for 5 or 10 years. In your case since Let's Encrypt is an official CA you do not need to copy the certificate at all. Simply comment out `teslaProxyCert` in `/etc/twcmanager/config.json`

> (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)')))

This is definitely a certificate issue, which you can fix by commenting out `teslaProxyCert` in `/etc/twcmanager/config.json`

> requests.exceptions.InvalidSchema: No connection adapters were found for 'my.secretdomain.com:4430/api/1/vehicles'

This looks like you made a typo in your `teslaProxy` setting, make sure it starts with `https://`

```
        "teslaProxy": "https://my.secretdomain.com:4430",
```

**Reply by alexeiw123** [reply]
> > Very interesting. My car is 2019 so almost certainly yes for 1. (and teslamate is still working fine on legacy API for now)
> 
> So stop trying to get the proxy to work since it looks you don't need it. Simply only define `teslaApiClientID` and use the refresh token from the Fleet API (HA integration). Comment out `teslaProxy` and `teslaProxyCert` in `/etc/twcmanager/config.json`

I think I'm missing something basic here, `teslaApiClientID` doesn't go hand in hand with the fleet API tool?

I've tried running it in the way you are suggesting, and while the other day it was letting me control charge rate, I'm now just being prompted to get TeslaAPI credentials again.

> It is very uncommon to use a public domain and an officially signed certificate for your Tesla HTTP Proxy. Generally the proxy is run on localhost with a self signed certificate that is valid for 5 or 10 years. In your case since Let's Encrypt is an official CA you do not need to copy the certificate at all. Simply comment out `teslaProxyCert` in `/etc/twcmanager/config.json`

The way this is set up is required for HA installs that have external access using cloudflared (the HA add-on to use a cloudflare tunnel). 
I'm now not directing to a certificate.

> This looks like you made a typo in your `teslaProxy` setting, make sure it starts with `https://`
> 
> ```
>         "teslaProxy": "https://my.secretdomain.com:4430",
> ```

Have fixed this part of it.

Now I'm struggling to move forward because something that I've done has caused any meaningful logging to stop. I can make config changes and restart the service but the logs always stop immediately, where previously I was able to see any errors (like the SSL error above).

```
journalctl -f
-- Logs begin at Sat 2024-04-20 17:17:02 AEST. --
Apr 22 07:57:27 TWCManager systemd[1]: twcmanager.service: Succeeded.
Apr 22 07:57:27 TWCManager systemd[1]: Stopped Tesla Wall Charger Manager.
Apr 22 07:57:27 TWCManager sudo[30286]: pam_unix(sudo:session): session closed for user root
Apr 22 07:58:40 TWCManager sudo[30328]:       pi : TTY=pts/0 ; PWD=/etc/twcmanager ; USER=root ; COMMAND=/usr/bin/nano config.json
Apr 22 07:58:40 TWCManager sudo[30328]: pam_unix(sudo:session): session opened for user root by pi(uid=0)
Apr 22 08:00:32 TWCManager sudo[30328]: pam_unix(sudo:session): session closed for user root
Apr 22 08:00:37 TWCManager sudo[30331]:       pi : TTY=pts/0 ; PWD=/etc/twcmanager ; USER=root ; COMMAND=/usr/sbin/service twcmanager start
Apr 22 08:00:37 TWCManager sudo[30331]: pam_unix(sudo:session): session opened for user root by pi(uid=0)
Apr 22 08:00:48 TWCManager systemd[1]: Started Tesla Wall Charger Manager.
Apr 22 08:00:48 TWCManager sudo[30331]: pam_unix(sudo:session): session closed for user root
```

**Reply by RichieB2B** [reply]
> I think I'm missing something basic here, `teslaApiClientID` doesn't go hand in hand with the fleet API tool?

`teslaApiClientID` goes hand in hand with the Tesla Fleet API app enrolment. If you generate a refresh token using the Fleet API app enrolment you need to set the correct `teslaApiClientID` to be able to actually refresh the token. If you use the (legacy) Owner API to get a refresh token, you do not need to set `teslaApiClientID`.

> I've tried running it in the way you are suggesting, and while the other day it was letting me control charge rate, I'm now just being prompted to get TeslaAPI credentials again.

This means your refresh token is no longer valid. Generate a new one.

> Now I'm struggling to move forward because something that I've done has caused any meaningful logging to stop.

Check the `logging` section of `/etc/twcmanager/config.json` It looks like you had `Console` enabled but now it is not. I believe the default is `FileLogger`, check the `path` in that section for the logfile location.

**Reply by alexeiw123** [reply]
> If you use the (legacy) Owner API to get a refresh token, you do not need to set `teslaApiClientID`

I guess trying to steer away from the complications I have getting the fleet API working correctly in TWCManager, if I can still use legacy then that will be a lot easier. What's the method to get legacy access and refresh tokens now? If I use the TWCManager UI link, it populates with values, gives a success message then goes straight back to prompting for the tokens again.

> This means your refresh token is no longer valid. Generate a new one.

Got this working again thanks, just no start/stop etc.

**Reply by RichieB2B** [reply]
For the Owner API tokens I used the TeslaFi iOS app but it is no longer available. 

I found this project that looks promising: https://github.com/adriankumpf/tesla_auth

**Reply by alexeiw123** [reply]
> I found this project that looks promising: https://github.com/adriankumpf/tesla_auth

This does seem to have worked, at a glance anyway. TWCManager has accepted it and I'm able to start/stop charge sessions again. Hopefully it remains authorised! No need for fleetAPI if legacy is working again for me :)

Thanks for all your help. I'm sure it will help someone else at some point too.

**Reply by RichieB2B** [reply]
You're welcome. Make sure to comment out `teslaApiClientID` in your `config.json` otherwise refreshing the Owner API token will probably fail.

**Reply by alexeiw123** [reply]
> You're welcome. Make sure to comment out `teslaApiClientID` in your `config.json` otherwise refreshing the Owner API token will probably fail.

Yes I've done that. I've noticed that it's not very persistent - auth seems to remain as long as TWC is powered up but if I reboot TWCManager then I need to generate new tokens again. Not sure that is expected behaviour - it certainly wasn't like that 6 months+ ago before I started getting difficulties with Tesla API.

**Reply by RichieB2B** [reply]
That is not supposed to happen. Check the modification time and permission of `/etc/twcmanager/settings.json` It should be writable by the `twcmanager` user.

**Reply by alexeiw123** [reply]
```
pi@TWCManager:/etc/twcmanager $ stat settings.json
  File: settings.json
  Size: 9003            Blocks: 24         IO Block: 4096   regular file
Device: b302h/45826d    Inode: 130187      Links: 1
Access: (0755/-rwxr-xr-x)  Uid: ( 1001/twcmanager)   Gid: ( 1001/twcmanager)
Access: 2024-02-05 14:16:34.871468161 +1100
Modify: 2024-04-30 08:30:41.974816083 +1000
Change: 2024-04-30 08:30:41.974816083 +1000
 Birth: -
```
I think this indicates twcmanager as owner and read/write and execute permissions? Also that it has been updated recently.
Things have been a little odd in my install for a while... for example:
![image](https://github.com/ngardiner/TWCManager/assets/77809797/f17cba5c-1a3b-49bf-9153-5c5c710d666e)

Another possibility is that DNS sinking network ad-blocking is preventing some token refresh request?

**Reply by RichieB2B** [reply]
That `stat` looks ok. Check if you actually have a refresh token with:
```
jq '.carApiRefreshToken,.carApiTokenExpireTime' /etc/twcmanager/settings.json
```
You can convert the expiry time (the last line of that output) on https://www.epochconverter.com
Install `jq` with `apt install jq` if you do not have it yet.

**Reply by alexeiw123** [reply]
Yep, that returned a refresh token,
The time returned is about 6 hours ahead of now

**Reply by RichieB2B** [reply]
Check again in 6+ hours.. seems to be working as expected.
